FM2-562: Fix pagination to address slow running queries

rkorytkowski opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'FM2-8: Implement the Person Resource' -->
<!--- 'FM2-JiraIssueNumber: JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/FM2-

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

ibacher reviewed (approved):
Thanks @rkorytkowski! This looks great! My only real comment is that, for w/e reason, the OpenMRS coding standard is to never use `*` imports.

rkorytkowski commented:
This is an initial pass. There's still one failing tests that I need to fix. I have trouble understanding the logic behind the method that is being tested so I'd appreciate some explanation.

The fix is to use DB pagination instead of paginating on the java side. Alongside I changed the code to run only 1 query and not two queries to get results in most cases as there's no need to first get IDs and then actual results. I also removed distinct filtering on results in places where it is not needed and made the distinct work on the DB side. The count also uses proper select count(*) and not size() on the results.

rkorytkowski commented:
> Thanks @rkorytkowski! This looks great! My only real comment is that, for w/e reason, the OpenMRS coding standard is to never use `*` imports.

I'll fix that. Apparently my IDE is not configured properly.

rkorytkowski commented:
@ibacher Thanks! Fixed the imports and the failing test. There's one thing with the new implementation for lastn obs feature that the count of results may be off in a rare condition. I commented in code. It feels such rare that I would not sacrifice performance to support it. I'm also not sure though if we expose that count anywhere...

ibacher commented on api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirObservationDaoImpl.java:
```diff
@@ -59,26 +55,92 @@ public class FhirObservationDaoImpl extends BaseFhirDao<Obs> implements FhirObse
+			List<Obs> results = new ArrayList<>();
+			int firstResult = 0;
+			int maxGroupCount = getMaxParameter(theParams);
+			final int BATCH_SIZE = 3;
```
I'm not sure I quite understand why we're using the constant 3 here?

rkorytkowski replied:
I was testing the algorithm... I forgot to revert it back to 100. Good eye ;)

ibacher commented:
@rkorytkowski The `count` does end up exposed as the `total` field in the search results. That said, I don't think that having that exactly right is super critical...

rkorytkowski commented:
> @rkorytkowski The `count` does end up exposed as the `total` field in the search results. That said, I don't think that having that exactly right is super critical...

Actually, I misunderstood the algorithm a bit, which is why the integration tests were failing and it may happen more commonly. I made the adjustments to the count so it's always accurate.

I also noticed that getting counts was called twice during a request so decided to cache it for the lifespan of the request.

ibacher commented:
> I also noticed that getting counts was called twice during a request so decided to cache it for the lifespan of the request.

Ooof... I hadn't that before. Thanks that should be good. I'm happy for you to merge this in whenever you're comfortable doing so, though it would be nice to hear back from the OHRI team whether this helps.
